src/lib/png/png.c

```diff
@@ -0,0 +1,1129 @@
+/**
+ * @file
+ * @brief lib for read and draw PNG
+ *
+ * @date May 10, 2019
+ * @author Filipp Chubukov
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+
+#include <linux/fb.h>
+
+#define PLT 1 /* pallete */
+#define TRCL 2 /* truecolor */
+#define TRCLA 3 /* truecolor + alpha */
+
+/* 256 literals, end code, length codes + 2 unused codes */
+#define NUMBER_OF_CODE_SYMBOLS 288
+
+/* the distance codes have their own symbols, 30 used, 2 unused */
+#define NUMBER_OF_DISTANCE_SYMBOLS 32
+
+/* lengths codes */
+#define NUMBER_OF_LENGTH_CODES 19
+#define FIRST_LENGTH_INDEX 257
+#define LAST_LENGTH_INDEX 285
+
+
+/*
+the base of distances(the bits of distance
+codes appear after length codes and use their own huffman tree)
+*/
+const int DISTANCE_BASE[30]
+	= {1, 2, 3, 4, 5, 7, 9, 13, 17, 25, 33, 49, 65, 97, 129, 193, 257, 385, 513,
+		769, 1025, 1537, 2049, 3073, 4097, 6145, 8193, 12289, 16385, 24577};
+
+/* addit bits of distances (added to base)*/
+const int DISTANCE_ADDITIONAL[30]
+	= {0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8,
+		8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13};
+
+/* represented by codes 257-285 */
+const int LENGTH_BASE[29]
+	= {3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 15, 17, 19, 23, 27, 31, 35, 43, 51, 59,
+		67, 83, 99, 115, 131, 163, 195, 227, 258};
+
+/* addit bits with codes 257-285 (added to base length) */
+const int LENGTH_ADDITIONAL[29]
+	= {0, 0, 0, 0, 0, 0, 0,  0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
+		4, 4, 4, 4, 5, 5, 5, 5, 0};
+/*
+the order in which "code length alphabet code lengths" are stored, out of this
+the huffman tree of the dynamic huffman tree lengths is generated
+*/
+const int CLCL_ORDER[NUMBER_OF_LENGTH_CODES]
+	= {16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15};
+
+#define READBIT(bitpointer, bitstream) ((bitstream[bitpointer >> 3] >> (bitpointer & 0x7)) & (unsigned char)1)
+
+/* def Huffmantree struct and functions to work with it */
+typedef struct HuffmanTree {
+	int *tree2d;
+	int *tree1d;
+	int *lengths; /* lengths of the codes(tree1d) */
+	int maxbitlen; /* max bits that single code can get */
+	int num_of_codes; /* number of symbols in the alph */
+} HuffmanTree;
+
+void HuffmanTree_init(HuffmanTree* tree) {
+	tree->tree2d = 0;
+	tree->tree1d = 0;
+	tree->lengths = 0;
+}
+
+void HuffmanTree_cleanup(HuffmanTree* tree) {
+	free(tree->tree2d);
+	free(tree->tree1d);
+	free(tree->lengths);
+}
+
+/*Two functions to convert string of bytes to int */
+int mypow(int digit,int degree) {
+	int i, res = 1;
+	for (i = 0; i < degree; i++) {
+		res = res * digit;
+	}
+	return res;
+}
+
+int convert(unsigned char *bytes, int len) {
```

Comment:
That's a bad name for a global function

Reply:
rework png and add static and global functions